### ISSUE TITLE
Symbolic links can change script behavior due to lack of invocation information

### ISSUE THREAD

**Issue by blightbow** [PowerShell/PowerShell#21402]
### Prerequisites

- [X] Write a descriptive title.
- [X] Make sure you are able to repro it on the [latest released version](https://github.com/PowerShell/PowerShell/releases)
- [X] Search the existing issues.
- [X] Refer to the [FAQ](https://github.com/PowerShell/PowerShell/blob/master/docs/FAQ.md).
- [X] Refer to [Differences between Windows PowerShell 5.1 and PowerShell](https://learn.microsoft.com/powershell/scripting/whats-new/differences-from-windows-powershell).

### Steps to reproduce

This is a symptom #4217, which was closed due to inactivity. Per "reopen a new issue and reference this issue if this is still a blocker for you", I am submitting this bug report as an example of why this issue is not simply cosmetic and needs to be fixed.

On UNIX based operating systems it is common to create user friendly aliases to commands using symbolic links. Symbolic links **should not** result in functional changes to script behavior, barring the case where the script itself is evaluating the name of the command used to invoke it (similar to what `busybox` does). The opposite applies here: a script cannot even test for its own name when the bug presents itself.

### Expected behavior

```console
Create a file named test.ps1 with the following contents:


#!/usr/bin/env pwsh
$PSCommandPath
$PSScriptRoot
$MyInvocation
```

Make the script executable and create a symbolic link to test.ps1 that does not contain a suffix of ".ps1":

```
chmod 755 test.ps1
ln -s test.ps1 test
```

The output of the script should be identical regardless of how it is invoked. Invoking the script on its own yields the expected result:

```
PS /tmp> ./test.ps1
/tmp/test.ps1
/tmp

MyCommand             : test.ps1
BoundParameters       : {}
UnboundArguments      : {}
ScriptLineNumber      : 1
OffsetInLine          : 1
HistoryId             : 5
ScriptName            :
Line                  : ./test.ps1
Statement             : ./test.ps1
PositionMessage       : At line:1 char:1
                        + ./test.ps1
                        + ~~~~~~~~~~
PSScriptRoot          :
PSCommandPath         :
InvocationName        : ./test.ps1
PipelineLength        : 1
PipelinePosition      : 1
ExpectingInput        : False
CommandOrigin         : Runspace
DisplayScriptPosition :
```
```


### Actual behavior

```console
The output changes because the absence of a ".ps1" extension resulted in the invocation information not being populated. (#4217)


PS /tmp> ./test



MyCommand             : #!/usr/bin/env pwsh
                        $PSCommandPath
                        $PSScriptRoot
                        $MyInvocation

BoundParameters       : {}
UnboundArguments      : {}
ScriptLineNumber      : 0
OffsetInLine          : 0
HistoryId             : 1
ScriptName            :
Line                  :
Statement             :
PositionMessage       :
PSScriptRoot          :
PSCommandPath         :
InvocationName        :
PipelineLength        : 2
PipelinePosition      : 1
ExpectingInput        : False
CommandOrigin         : Runspace
DisplayScriptPosition :
```
```


### Error details

_No response_

### Environment data

```powershell
Name                           Value
----                           -----
PSVersion                      7.4.1
PSEdition                      Core
GitCommitId                    7.4.1
OS                             Ubuntu 22.04.4 LTS
Platform                       Unix
PSCompatibleVersions           {1.0, 2.0, 3.0, 4.0…}
PSRemotingProtocolVersion      2.3
SerializationVersion           1.1.0.1
WSManStackVersion              3.0
```

**Comment by rhubarb-geek-nz** [comment]
Try running both from bash instead of PowerShell. What you will see is that when run with the ps1 extension PowerShell interprets it directly and ignores the `#!/usr/bin/env pwsh`. Without the ps1 extension PowerShell will simply use the fork/exec to run it.

So when run from PowerShell with the ps1 extension much more is known than running using the "#!/usr/bin/env pwsh" executable prefix.

On another note, remove the symbolic link so you only have "test.ps1" and PowerShell will run ./test by looking for a ./test.ps1 and running that.

**Comment by rhubarb-geek-nz** [comment]
On a final note, PowerShell should not have to follow OS symbolic links to find out what something is referring to when executing a program and neither should it be reading an executable and trying to determine how it has been implemented, that is the job of the exec system call.

**Comment by blightbow** [comment]
To simplify analysis and drive the point home, I will provide additional examples that are executed from bash. I am also removing the /usr/bin/env abstraction in favor of an absolute path in the shebang line.

* `test2.ps1` - Same as test1.ps1, but an absolute path in the shebang line.
* `test2link` - A symbolic link to test2.
* `test3` - The contents are exactly the same as test2.ps1. The only difference is the name of the file.

```
# what shell is this running in?
$ readlink /proc/$$/exe
/usr/bin/bash

# create a symbolic link named test2link that points at test2.ps1
$ ln -s test2.ps1 test2link

# copy test2.ps1 to test3
$ cp test2.ps1 test3

# confirm that all three files produce the same sha1sum
$ sha1sum test2.ps1 test2link test3
49b77df919a70469a111ab644d0eea5650ef2b4b  test2.ps1
49b77df919a70469a111ab644d0eea5650ef2b4b  test2link
49b77df919a70469a111ab644d0eea5650ef2b4b  test3

# confirm the contents of test2.ps1 before we begin
$ cat test2.ps1
#!/opt/microsoft/powershell/7/pwsh
$PSCommandPath
$PSScriptRoot
$MyInvocation

$ ./test2.ps1
/tmp/test2.ps1
/tmp

MyCommand             : test2.ps1
BoundParameters       : {}
UnboundArguments      : {}
ScriptLineNumber      : 0
OffsetInLine          : 0
HistoryId             : 1
ScriptName            :
Line                  :
Statement             :
PositionMessage       :
PSScriptRoot          :
PSCommandPath         :
InvocationName        : /tmp/test2.ps1
PipelineLength        : 2
PipelinePosition      : 1
ExpectingInput        : False
CommandOrigin         : Runspace
DisplayScriptPosition :

$ ./test2link



MyCommand             : #!/opt/microsoft/powershell/7/pwsh
                        $PSCommandPath
                        $PSScriptRoot
                        $MyInvocation

BoundParameters       : {}
UnboundArguments      : {}
ScriptLineNumber      : 0
OffsetInLine          : 0
HistoryId             : 1
ScriptName            :
Line                  :
Statement             :
PositionMessage       :
PSScriptRoot          :
PSCommandPath         :
InvocationName        :
PipelineLength        : 2
PipelinePosition      : 1
ExpectingInput        : False
CommandOrigin         : Runspace
DisplayScriptPosition :

$ ./test3



MyCommand             : #!/opt/microsoft/powershell/7/pwsh
                        $PSCommandPath
                        $PSScriptRoot
                        $MyInvocation

BoundParameters       : {}
UnboundArguments      : {}
ScriptLineNumber      : 0
OffsetInLine          : 0
HistoryId             : 1
ScriptName            :
Line                  :
Statement             :
PositionMessage       :
PSScriptRoot          :
PSCommandPath         :
InvocationName        :
PipelineLength        : 2
PipelinePosition      : 1
ExpectingInput        : False
CommandOrigin         : Runspace
DisplayScriptPosition :

```

Illustration that other shells provide consistent invocation information:

```
$ cat bash.sh
#!/bin/bash
echo $0

$ ln -s bash.sh bashlink

$ chmod 755 bash.sh

$ sha1sum bash.sh bashlink
78fd4ed1622a581ca0a81676dbf83ce644ca85f8  bash.sh
78fd4ed1622a581ca0a81676dbf83ce644ca85f8  bashlink

$ ./bash.sh
./bash.sh

$ ./bashlink
./bashlink

## Debian's dash shell
$ cat dash.sh
#!/bin/dash
echo $0

$ ln -s dash.sh dashlink

$ chmod 755 dash.sh

$ sha1sum dash.sh dashlink
f0763356461e8a06757e20eff2fef3883dbf15d8  dash.sh
f0763356461e8a06757e20eff2fef3883dbf15d8  dashlink

$ ./dash.sh
./dash.sh

$ ./dashlink
./dashlink
```

**Comment by rhubarb-geek-nz** [comment]
From your tests I think it suggests that the behaviour of PowerShell depends on the ps1 extension rather than the symbolic link. If it is a ps1 extension then MyCommand is the path to the script, if not it is the full contents of the file. It literally does not know it is a script file so it just reads the contents as the command to run and tries to interpret it.

Notice that with ps1 extension you then have invocation name populated with the path to the script. Even with the ps1 extension the PSScriptRoot was not populated.

UNIX shells however do not care about extensions, Using *.sh is merely a convention, not a requirement. PowerShell comes from windows where the extension cmd, bat, exe, com is important in determining how to handle executable files.

Despite its name, PowerShell is not a UNIX shell, it is much more like Perl or Python and I suggest that consistency with PowerShell  across each platform is more important than trying to emulate behaviour from UNIX shells.

**Comment by blightbow** [comment]
> From your tests I think it suggests that the behaviour of PowerShell depends on the ps1 extension rather than the symbolic link.

Correct. As mentioned before this is a symptom of #4217, which describes this exact behavior. My intent was to highlight that this behavior could create functional differences in script behavior with little more than a symbolic link, which is concerning from UNIX administrator's perspective.

> Despite its name, PowerShell is not a UNIX shell, it is much more like Perl or Python and I suggest that consistency with PowerShell across each platform is more important than trying to emulate behaviour from UNIX shells.

I respectfully submit that this is a false equivalence. Your cited examples provide a consistent interface to invocation information. Of the shells and languages that have been discussed in these comments, Powershell is the only one I am aware of that masks the user facing interfaces to the name of the script depending on how the file is named. If you can provide a counter example I will happily concede the point, but my experience with Linux, Perl, and Python all suggest that this is not the case.

> UNIX shells however do not care about extensions, Using *.sh is merely a convention, not a requirement. PowerShell comes from windows where the extension cmd, bat, exe, com is important in determining how to handle executable files.

I would not certainly expect a Python script to directly execute on a Windows OS if it lacked a .py suffix. Likewise, I would not expect a Powershell script to directly execute on a UNIX based OS if I did not either modify it to include a shebang line and apply an executable bit. Prefacing the Python or Powershell script with the path to the intended interpreter allows both scripts to run without modification.

The difference between these is that Powershell simply makes no attempt to populate the invocation information if the suffix does not end in ps1. Python does not do this with its extension. Perl does not do this with its extension. It is happening because of a design choice inside of the interpreter, not because it has to.

If this is the design choice of the Powershell developers and they intend to be the sole interpreter that deviates from common UNIX conventions, this is fine. _Questionable_, but fine. I will not belabor the point, but I do want the community to understand that this is the reason why this behavior persists and not because there is a lack of understanding as to why this should change. (as per #4217)

**Comment by mklement0** [comment]
@rhubarb-geek-nz, to add to @blightbow's comment:

> The behaviour of PowerShell depends on the ps1 extension

Indeed - and it _shouldn't_.

> It literally does not know it is a script file so it just reads the contents as the command to run and tries to interpret it.

It _does_ know that it is a script file, because a shebang line is translated _by the system_ into passing the _file path_ to the interpreter specified in the shebang line.

In short: The problem is the lack of "self-awareness" when an _extension-less_ script is executed  - whether explicitly or via a shebang line -  and this lack is _clearly a bug_.

**Comment by rhubarb-geek-nz** [comment]
> 
> > The behaviour of PowerShell depends on the ps1 extension
> 
> Indeed - and it _shouldn't_.

Within PowerShell itself, if you run a script with a ps1 extension it is run within the same process. Without ps1 it has no idea what kind of executable it is so uses fork/exec, this will run the script or executable in a separate process. This is existing behaviour.

**Comment by blightbow** [comment]
> Within PowerShell itself, if you run a script with a ps1 extension it is run within the same process. Without ps1 it has no idea what kind of executable it is so uses fork/exec, this will run the script or executable in a separate process. This is existing behaviour.
>
> In comparison, UNIX shells always run a new script in a separate process unless you dot-source them.

This sounds like a consequence of an interpreter that needs to fork its behavior to handle multiple formats of text files without standardizing on a default one. It also sounds like behavior that could be solved with a simple commandline switch, or (ironically) a symlink to the pwsh executable with different behavior. :)

I understand that there will be significant inertia to implementing these solutions, but it also stands to reason that Powershell appears to be the only modern text interpreter with this design problem that I've seen on a UNIX OS. (again, I am happy to concede to counter-examples) Occupying such a unique design constraint that differentiates Powershell from other interpreters does in fact suggest that an alternative startup mode is warranted.

**Comment by mklement0** [comment]
@rhubarb-geek-nz, here's a self-contained example that demonstrates that PowerShell _does_ know what file path it was invoked with via a shebang line (by design resulting in  a _separate_ PowerShell (child) process, but that  is _beside the point_); as noted in my previous comment this is _system_-level behavior, unrelated to PowerShell:

```powershell
@'
#!/usr/bin/env pwsh
[pscustomobject] @{
  CommandLine = [Environment]::GetCommandLineArgs()[1]
  PSCommandPath = $PSCommandPath
}
'@ > foo
chmod a+x foo
./foo  # Effectively the same as: pwsh -File "$PWD/foo"
```

There is _no_ justification for PowerShell not reflecting the given script-file path in `$PSCommandPath` / `$PSScriptRoot` / `$MyInvocation` simply because the file path _happens not to have a `.ps1` extension_.

**Comment by rhubarb-geek-nz** [comment]
> There is _no_ justification for PowerShell not reflecting the given script-file path in `$PSCommandPath` / `$PSScriptRoot` / `$MyInvocation` simply because the file path _happens not to have a `.ps1` extension_.

The current behaviour does support passing /dev/fd/4 or similar as the command line argument to pwsh eg to support remote servers for example. It's current behaviour is it isn't a script file but I can try and read it, either a socket or a pipe.

**Comment by mklement0** [comment]
To be clear: The following aspects are _incidental_ to the issue at hand:

* That files with extension `.ps1` files run _in-process_ by design - whether or not they have as shebang line.

* Whether or not a shebang line-based script that happens to use PowerShell as its interpreter is invoked _via a symlink_ (if you create a `.ps1` symlink that targets an extension-less shebang line-based script, things work as expected, with the _symlink_'s path being reflected in `$PSCommandPath`)

**Comment by mklement0** [comment]
@rhubarb-geek-nz:

> support passing /dev/fd/4 or similar 

That's because its default CLI parameter is `-File`, expecting a path to a _file_ as the argument whose content is PowerShell code. Of course this means that you may also pass  that path of _special files_ such as `/dev/fd/4` (from Bash, try  `pwsh <(echo '[Environment]::CommandLine')`) - again, this has nothing to do with the issue at hand.

**Comment by 237dmitry** [comment]
This is not problem of symbolic link itself but the name of file. With ps1 extension scripts and theirsymlinks working in pwsh and bash as expected:

```powershell
$ dir aaa*,bbb*,ccc*,ddd*,eee*

    Directory: /tmp

UnixMode     Num   UID   GID      LastWriteTime    Size Name
--------     ---   ---   ---      -------------    ---- ----
-rwxr-xr-x   755  1000  1000   02.04.2024 12:56      63 aaa.ps1                       # working
-rwxr-xr-x   755  1000  1000   02.04.2024 12:56      63 bbb                           # not working
lrwxrwxrwx   755  1000  1000   02.04.2024 12:58       9 ccc -> ./aaa.ps1              # not working
lrwxrwxrwx   755  1000  1000   02.04.2024 12:59       9 ddd.ps1 -> ./aaa.ps1          # working
lrwxrwxrwx   755  1000  1000   02.04.2024 13:07       3 eee.ps1 -> bbb                # working

```

**Comment by mklement0** [comment]
@rhubarb-geek-nz, the - to me problematic - pseudo-interactive behavior with `-` (stdin, both with `-File` and `-Command`) is yet another issue, unrelated to the bug at hand - see #3223

> On Posix PowerShell does attempt to run other types of files by reading their content and interpreting as a series of PowerShell statements.

That's pretty much how all shells and interpreters work (POSIX-compatible shells, Python, Ruby, Perl, ...); PowerShell _on Windows_, with its requirement that the file have extension `.ps1`, is the exception. Given PowerShell's need to support shebang line-based scripts on Unix-like platforms, this restriction had to be lifted there (along with switching the default CLI parameter from `-Command` to `-File`).

While it's good to know _workarounds_ (suggestion: `%~dp0test.ps1` -> `"%~dpn0.ps1"`) and have background information, I suggest we stay focused on the bug at hand, which is fully described [above](https://github.com/PowerShell/PowerShell/issues/21402#issuecomment-2030975504).

**Comment by rhubarb-geek-nz** [comment]
> That's pretty much how all shells and interpreters work

I meant absolutely literally, if the file has a ps1 extension then MyCommand contains the filename, if not then it contains the file contents.

>While it's good to know workarounds 

It certainly is, it stops issues being treated as show stoppers, it unblocks people from working now and gives you a backward compatible solution and the bug fix may be months or years away if ever.

**Comment by mklement0** [comment]
> if the file has a ps1 extension then MyCommand contains the filename, if not then it contains the file contents.

Yes, PowerShell _choosing_ do to that - inappropriately - appears to be the bug.

Agreed re workarounds - if succinctly presented, while acknowledging the bug and avoiding distractions.

**Comment by rkeithhill** [comment]
Perhaps related: https://github.com/PowerShell/PowerShell/issues/4217

**Comment by rkeithhill** [comment]
Also related: https://github.com/PowerShell/PowerShell/issues/16480#issuecomment-971897032

**Comment by rkeithhill** [comment]
The Engine WG reviewed this today.  We agree with @mklement0 that this issue is really about how PowerShell handles extensionless scripts on *nix.  There's agreement on having `$PSCommandPath` & `$PSScriptRoot` be populated in this case.  IOW what reason is there not to populate these built-in variables?  There are some security concerns around this issue that need to be investigated.
